File: 백준/BinarySearch/2805.js
const fs = require("fs");

const inputs = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const [n, m] = inputs[0].split(" ").map((e) => Number(e));
const trees = inputs[1].split(" ").map((e) => Number(e));

let left = 0;
let right = Math.max(...trees);
let mid = Math.floor((left + right) / 2);

while (left <= right) {
  const cutted = trees.reduce((a, c) => {
    return a + Math.max(0, c - mid);
  }, 0);

  if (cutted >= m) {
    left = mid + 1;
  } else {
    right = mid - 1;
  }

  mid = Math.floor((left + right) / 2);
}

console.log(right);

File: 백준/BinarySearch/2805.py
import sys
import math

n, m = map(int, sys.stdin.readline().split())
trees = list(map(int, sys.stdin.readline().split()))



max_height = max(trees)
left = 0
right = max_height
mid = math.floor((left + right) / 2)

while left <= right:
    cnt = sum(map(lambda tree: max(0, tree - mid), trees))
   
    if cnt < m:
        right = mid - 1
    else:
        left = mid + 1
    
    mid = math.floor((left + right) / 2)

print(mid)